File: 2024/day13/2.ts
module.exports = {
    main: main
};

interface Vector2 {
    x: number,
    y: number
}

interface MachineData {
    buttonA: Vector2,
    buttonB: Vector2,
    prize: Vector2
}
export function main(input: string[]): any {
    const machines: MachineData[] = loadData(input);
    let coins = 0;
    for (let i = 0; i < machines.length; i++) {
        coins += playMachine(machines[i]);
    }

    return coins;
}

function loadData(input: string[]): MachineData[] {
    const machines: MachineData[] = [];
    for (let i = 0; i < input.length; i += 4) {
        var numberPattern = /\d+/g;
        const vectors: Vector2[] = [];

        for (let j = 0; j < 3; j++) {
            const numbers: string[] = input[i + j].match(numberPattern);
            vectors.push({ x: Number(numbers[0]), y: Number(numbers[1]) });
        }
        vectors[2].x += 10000000000000;
        vectors[2].y += 10000000000000;
        machines.push({ buttonA: vectors[0], buttonB: vectors[1], prize: vectors[2] });
    }

    return machines;
}

function playMachine(machine: MachineData): number {
    let coins = Infinity;
    let MaxX = machine.prize.x / machine.buttonA.x;
    let MaxY = machine.prize.y / machine.buttonA.y;
    console.log(MaxX);
    for (let nA = 0; nA <= MaxX; nA++) {
        console.log(`${nA} of ${MaxX}`);
        let nB = (machine.prize.x - machine.buttonA.x * nA) / machine.buttonB.x;
        if (!Number.isInteger(nB))
            continue;
        if (machine.buttonA.y * nA + machine.buttonB.y * nB == machine.prize.y) {
            const newCoins = 3 * nA + nB;
            coins = coins < newCoins ? coins : newCoins;
        }

    }
    if (coins == Infinity)
        return 0;
    return coins;
}

File: 2024/day13/2.py
from ortools.linear_solver import pywraplp
import re

def findSolution(goal, A, B):
    solver = pywraplp.Solver.CreateSolver('SCIP')
    nA = solver.IntVar(0, solver.infinity(), name= "A button")
    nB = solver.IntVar(0, solver.infinity(), name= "B button")
    obj = solver.Objective()
    obj.SetCoefficient(nA, 3)
    obj.SetCoefficient(nB, 1)
    obj.SetMinimization()

    xPos = solver.Constraint(goal["x"], goal["x"])
    xPos.SetCoefficient(nA, A["x"])
    xPos.SetCoefficient(nB, B["x"])

    yPos = solver.Constraint(goal["y"], goal["y"])
    yPos.SetCoefficient(nA, A["y"])
    yPos.SetCoefficient(nB, B["y"])

    result = solver.Solve()
    if result == solver.INFEASIBLE:
        print("infeasable")
        return 0
    elif result == solver.OPTIMAL:
        print(nA.solution_value(), nB.solution_value())
        return obj.Value()
    

def handleInput(path):
    # Define a number pattern to match (similar to your numberPattern in TypeScript)
    number_pattern = r'\d+'  # This will match sequences of digits in the text
    machines = []
    # Open the file in read mode
    with open(path, 'r') as file:
        lines = file.readlines()  # Read all lines into a list
        
        # Process lines in chunks of 4
        for i in range(0, len(lines), 4):
            # Iterate through the first 3 lines in each chunk (if available)
            data = []
            for j in range(min(3, len(lines) - i)):  # Ensure we don't go out of bounds
                line = lines[i + j].strip()  # Get the current line, strip any extra whitespace
                
                # Find numbers in the line using the regular expression
                numbers = re.findall(number_pattern, line)
                data.append({"x": int(numbers[0]), "y": int(numbers[1])})
            
            data[2]["x"] += 10000000000000
            data[2]["y"] += 10000000000000
            machines.append(data)
    return machines    


machines = handleInput("build/day13/input.txt")
score = 0
for machine in machines:
    score += findSolution(machine[2], machine[0], machine[1])

print(score)